File: src/MewUI.Preview/PreviewProtocol.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Aprillz.MewUI.Preview;

/// <summary>
/// mew-preview v1 wire contract: 4-byte little-endian total length, 4-byte message type id,
/// 4-byte JSON length, UTF-8 JSON body, then an optional binary payload (Frame pixels).
/// Unknown message type ids are ignored for minor-version forward compatibility.
/// </summary>
internal static class PreviewProtocol
{
    internal const int PROTOCOL_MAJOR = 1;
    internal const int PROTOCOL_MINOR = 2;

    // Sanity cap for inbound messages; Frame (outbound) is bounded by the negotiated viewport.
    internal const int MAX_MESSAGE_BYTES = 64 * 1024 * 1024;

    internal const int HELLO = 1;
    internal const int SESSION_STARTED = 2;
    internal const int CLIENT_INFO = 3;
    internal const int PREVIEW_TARGETS = 4;
    internal const int SELECT_TARGET = 5;
    internal const int VIEWPORT_CHANGED = 6;
    internal const int FRAME = 7;
    internal const int FRAME_ACK = 8;
    internal const int STATUS = 9;
    internal const int REFRESH_TARGET = 10;
    internal const int SESSION_REJECTED = 11;
    internal const int SET_THEME = 12;
    internal const int POINTER_MOVED = 13;
    internal const int POINTER_PRESSED = 14;
    internal const int POINTER_RELEASED = 15;
    internal const int SCROLL = 16;
    internal const int KEY = 17;
    internal const int TEXT_INPUT = 18;
}

internal sealed class HelloMessage
{
    public int ProtocolMajor { get; set; }
    public int ProtocolMinor { get; set; }
    public string Token { get; set; } = string.Empty;
    public string[] Capabilities { get; set; } = [];
}

internal sealed class SessionStartedMessage
{
    public string SessionId { get; set; } = string.Empty;
    public string FrameworkVersion { get; set; } = string.Empty;
    public int ProtocolMajor { get; set; }
    public int ProtocolMinor { get; set; }
    public string[] Capabilities { get; set; } = [];
}

internal sealed class SessionRejectedMessage
{
    public string Reason { get; set; } = string.Empty;
}

internal sealed class ClientInfoMessage
{
    public double Dpi { get; set; }
    public double ViewportWidth { get; set; }
    public double ViewportHeight { get; set; }
}

internal sealed class PreviewTargetInfo
{
    public string Id { get; set; } = string.Empty;
    public string DisplayName { get; set; } = string.Empty;
    public string Kind { get; set; } = string.Empty;
    public string? SourcePath { get; set; }
    public int? SourceLine { get; set; }
    public bool Available { get; set; } = true;
    public string? Reason { get; set; }
}

internal sealed class PreviewTargetsMessage
{
    public PreviewTargetInfo[] Targets { get; set; } = [];
    public string ActiveId { get; set; } = string.Empty;
}

internal sealed class SelectTargetMessage
{
    public string Id { get; set; } = string.Empty;
}

internal sealed class ViewportChangedMessage
{
    public double Width { get; set; }
    public double Height { get; set; }
    public double Dpi { get; set; }
}

internal sealed class FrameMessage
{
    public long Seq { get; set; }
    public int Width { get; set; }
    public int Height { get; set; }
    public int Stride { get; set; }
    public string Format { get; set; } = "bgra8888";
    public double DpiScale { get; set; }
}

internal sealed class FrameAckMessage
{
    public long Seq { get; set; }
}

internal sealed class StatusMessage
{
    public string Message { get; set; } = string.Empty;
    public bool IsBuilding { get; set; }
    public bool HasError { get; set; }
    public string? UpdateKind { get; set; }
    public string? ExceptionDetail { get; set; }
    public string? ThemeMode { get; set; }
}

internal sealed class SetThemeMessage
{
    // "light" | "dark" | "system"
    public string Mode { get; set; } = string.Empty;
}

/// <summary>
/// Pointer state in window DIPs. Button uses the W3C convention (0 left, 1 middle, 2 right);
/// Buttons is the W3C bitmask (1 left, 2 right, 4 middle) after the event; Modifiers matches
/// the framework's ModifierKeys flags.
/// </summary>
internal sealed class PointerMessage
{
    public double X { get; set; }
    public double Y { get; set; }
    public int Button { get; set; }
    public int Buttons { get; set; }
    public int ClickCount { get; set; } = 1;
    public int Modifiers { get; set; }
}

internal sealed class ScrollMessage
{
    public double X { get; set; }
    public double Y { get; set; }
    // Notches: +Y = scroll-up intent, +X = scroll-left intent (framework wheel convention).
    public double DeltaX { get; set; }
    public double DeltaY { get; set; }
    public int Modifiers { get; set; }
}

internal sealed class KeyMessage
{
    // W3C KeyboardEvent.code (physical key), e.g. "KeyA", "ArrowLeft".
    public string Code { get; set; } = string.Empty;
    public bool IsDown { get; set; }
    public int Modifiers { get; set; }
}

internal sealed class TextInputMessage
{
    public string Text { get; set; } = string.Empty;
}

[JsonSourceGenerationOptions(PropertyNamingPolicy = JsonKnownNamingPolicy.CamelCase)]
[JsonSerializable(typeof(HelloMessage))]
[JsonSerializable(typeof(SessionStartedMessage))]
[JsonSerializable(typeof(SessionRejectedMessage))]
[JsonSerializable(typeof(ClientInfoMessage))]
[JsonSerializable(typeof(PreviewTargetsMessage))]
[JsonSerializable(typeof(SelectTargetMessage))]
[JsonSerializable(typeof(ViewportChangedMessage))]
[JsonSerializable(typeof(FrameMessage))]
[JsonSerializable(typeof(FrameAckMessage))]
[JsonSerializable(typeof(StatusMessage))]
[JsonSerializable(typeof(SetThemeMessage))]
[JsonSerializable(typeof(PointerMessage))]
[JsonSerializable(typeof(ScrollMessage))]
[JsonSerializable(typeof(KeyMessage))]
[JsonSerializable(typeof(TextInputMessage))]
internal sealed partial class PreviewJsonContext : JsonSerializerContext
{
}

File: tools/rider-mewui/src/main/kotlin/com/aprillz/mewui/preview/PreviewProtocol.kt
// mew-preview v1 wire contract, editor side (see MewUI/agent/preview-tooling/plan.md 4.3):
// 4-byte LE total length (excluding itself), 4-byte type id, 4-byte JSON length,
// UTF-8 JSON body, optional trailing binary payload (Frame pixels).
package com.aprillz.mewui.preview

import com.google.gson.Gson
import com.google.gson.JsonObject
import java.io.ByteArrayOutputStream

object PreviewProtocol {
    const val PROTOCOL_MAJOR = 1
    const val PROTOCOL_MINOR = 2
    const val MAX_MESSAGE_BYTES = 64 * 1024 * 1024

    const val HELLO = 1
    const val SESSION_STARTED = 2
    const val CLIENT_INFO = 3
    const val PREVIEW_TARGETS = 4
    const val SELECT_TARGET = 5
    const val VIEWPORT_CHANGED = 6
    const val FRAME = 7
    const val FRAME_ACK = 8
    const val STATUS = 9
    const val REFRESH_TARGET = 10
    const val SESSION_REJECTED = 11
    const val SET_THEME = 12
    const val POINTER_MOVED = 13
    const val POINTER_PRESSED = 14
    const val POINTER_RELEASED = 15
    const val SCROLL = 16
    const val KEY = 17
    const val TEXT_INPUT = 18

    val gson = Gson()

    fun encode(typeId: Int, body: Any): ByteArray {
        val json = gson.toJson(body).toByteArray(Charsets.UTF_8)
        val message = ByteArray(12 + json.size)
        writeInt(message, 0, json.size + 8)
        writeInt(message, 4, typeId)
        writeInt(message, 8, json.size)
        json.copyInto(message, 12)
        return message
    }

    private fun writeInt(buffer: ByteArray, offset: Int, value: Int) {
        buffer[offset] = value.toByte()
        buffer[offset + 1] = (value shr 8).toByte()
        buffer[offset + 2] = (value shr 16).toByte()
        buffer[offset + 3] = (value shr 24).toByte()
    }
}

class DecodedMessage(val typeId: Int, val json: JsonObject, val binary: ByteArray)

/** Incremental frame decoder: feed socket chunks, emits complete messages via the callback. */
class MessageDecoder(private val onMessage: (DecodedMessage) -> Unit) {
    private val buffer = ByteArrayOutputStream()
    private var consumed = 0

    fun push(chunk: ByteArray, count: Int) {
        buffer.write(chunk, 0, count)
        val data = buffer.toByteArray()
        var offset = consumed

        while (data.size - offset >= 12) {
            val totalLength = readInt(data, offset)
            require(totalLength in 8..PreviewProtocol.MAX_MESSAGE_BYTES) { "invalid message length $totalLength" }
            if (data.size - offset < 4 + totalLength) {
                break
            }

            val typeId = readInt(data, offset + 4)
            val jsonLength = readInt(data, offset + 8)
            val payloadLength = totalLength - 8
            require(jsonLength in 0..payloadLength) { "invalid json length $jsonLength" }

            val jsonText = String(data, offset + 12, jsonLength, Charsets.UTF_8)
            val binary = data.copyOfRange(offset + 12 + jsonLength, offset + 4 + totalLength)
            offset += 4 + totalLength

            onMessage(DecodedMessage(typeId, PreviewProtocol.gson.fromJson(jsonText, JsonObject::class.java), binary))
        }

        if (offset > 0) {
            val remaining = data.copyOfRange(offset, data.size)
            buffer.reset()
            buffer.write(remaining)
        }
        consumed = 0
    }

    private fun readInt(data: ByteArray, offset: Int): Int =
        (data[offset].toInt() and 0xff) or
            ((data[offset + 1].toInt() and 0xff) shl 8) or
            ((data[offset + 2].toInt() and 0xff) shl 16) or
            ((data[offset + 3].toInt() and 0xff) shl 24)
}

class PreviewTargetInfo(
    val id: String = "",
    val displayName: String = "",
    val kind: String = "",
    val sourcePath: String? = null,
    val sourceLine: Int? = null,
    val available: Boolean = true,
    val reason: String? = null,
)

class FrameHeader(
    val seq: Long = 0,
    val width: Int = 0,
    val height: Int = 0,
    val stride: Int = 0,
    val format: String = "bgra8888",
    val dpiScale: Double = 1.0,
)

class StatusInfo(
    val message: String = "",
    val isBuilding: Boolean = false,
    val hasError: Boolean = false,
    val updateKind: String? = null,
    val exceptionDetail: String? = null,
    val themeMode: String? = null,
)
